File: container_info.go
package main

import (
	"fmt"
	"gopkg.in/qml.v1"
)

type ContainerInfo struct {
	Id      string
	Name    string
	Image   string
	Created string
	Running bool
	Env     EnvironmentVariablesList
	Mounts  MountsList
	Ports   PortBindingsList
}

type EnvironmentVariablesList struct {
	list []EnvironmentVariable
	Len  int
}

func (self *EnvironmentVariablesList) Add(env []EnvironmentVariable) {
	self.list = append(self.list, env...)
	self.Len = len(self.list)
	qml.Changed(self, &self.Len)
}

func (self *EnvironmentVariablesList) Get(position int) *EnvironmentVariable {
	if position >= self.Len {
		fmt.Println("errrrr", position, self.Len)
	}
	return &self.list[position]
}

type EnvironmentVariable struct {
	Key   string
	Value string
}

type Mounts struct {
	Source      string
	Destination string
}

type MountsList struct {
	list []Mounts
	Len  int
}

func (self *MountsList) Add(env []Mounts) {
	self.list = append(self.list, env...)
	self.Len = len(self.list)
	qml.Changed(self, &self.Len)
}

func (self *MountsList) Get(position int) *Mounts {
	return &self.list[position]
}

type PortBinding struct {
	ContainerPort string
	HostPort      string
}

type PortBindingsList struct {
	list []PortBinding
	Len  int
}

func (self *PortBindingsList) Add(ports []PortBinding) {
	self.list = append(self.list, ports...)
	self.Len = len(self.list)
	qml.Changed(self, &self.Len)
}

func (self *PortBindingsList) Get(position int) *PortBinding {
	return &self.list[position]
}

File: container_info.py
from gi.repository import Gtk

from store import store


class ContainerInfo(Gtk.Bin):
    def __init__(self):
        Gtk.Bin.__init__(self)
        self.set_border_width(5)

        builder = Gtk.Builder()
        builder.add_from_file('./ui/container_info.glade')
        builder.connect_signals(self)

        self.add(builder.get_object('container_info'))

        # labels
        self.name_label = builder.get_object('name_label')
        self.image_label = builder.get_object('image_label')
        self.status_label = builder.get_object('status_label')

        # stores
        self.ev_store = builder.get_object('ev_store')
        """ :type : Gtk.ListStore """
        self.mounts_store = builder.get_object('mounts_store')
        """ :type : Gtk.ListStore """

    def update_info(self, container_id):
        container_info = store.get_container_info(container_id)
        self.name_label.set_label(
            '<span size="xx-large">' + container_info['Name'] + '</span>')
        self.image_label.set_label(container_info['Config']['Image'])
        for i in container_info:
            print(i, container_info[i], '\n')

        # status
        status = container_info['State']['Status'].upper()
        if status[0] == 'E':
            status = '<span color="red">'+status+'</span>'
        elif status[0] == 'R':
            status = '<span color="green">'+status+'</span>'
        self.status_label.set_label(status)

        # environment variables store
        self.ev_store.clear()
        env = container_info['Config']['Env']
        for i in env:
            i = i.split('=')
            self.ev_store.append(i)

        # mounts store
        self.mounts_store.clear()
        mounts = container_info['Mounts']
        for i in mounts:
            self.mounts_store.append([
                i['Destination'],
                i['Source'],
                i['Mode']
            ])
